string.go

```diff
@@ -0,0 +1,80 @@
+package surge
+
+import (
+	"reflect"
+	"strings"
+)
+
+func SizeHintString(v string) int {
+	return 2 + len(v)
+}
+
+func MarshalString(v string, buf []byte, rem int) ([]byte, int, error) {
+	buf, rem, err := MarshalU16(uint16(len(v)), buf, rem)
+	if err != nil {
+		return buf, rem, err
+	}
+	if len(buf) < len(v) || rem < len(v) {
+		return buf, rem, ErrUnexpectedEndOfBuffer
+	}
+	n := copy(buf, v)
+	return buf[n:], rem - n, nil
+}
+
+func UnmarshalString(v *string, buf []byte, rem int) ([]byte, int, error) {
+	strLen := uint16(0)
+	buf, rem, err := UnmarshalU16(&strLen, buf, rem)
+	if err != nil {
+		return buf, rem, err
+	}
+	n := int(strLen)
+	if n < 0 {
+		return buf, rem, ErrLengthOverflow
+	}
+	if len(buf) < n || rem < n {
+		return buf, rem, ErrUnexpectedEndOfBuffer
+	}
+	b := strings.Builder{}
+	if _, err := b.Write(buf[:n]); err != nil {
+		return buf[:n], rem - n, err
+	}
+	*v = b.String()
+	return buf[n:], rem - n, nil
+}
+
+func sizeHintReflectedString(v reflect.Value) int {
```

Comment:
Why do we have these reflected versions of the functions if we have the typed ones above?

Reply:
because we need them for the untyped entry functions `SizeHint`, `Marshal`, and `Unmarshal`, which cannot case their `interface{}` arguments to concrete types (because we check kindedness, not types, so that we can implicitly support definitions like `type MyInt64 int64`)

Reply:
@ross-pure Because we need them for the untyped entry functions `SizeHint`, `Marshal`, and `Unmarshal`. These functions cannot cast their `v interface{}` arguments to concrete types, because switch on kinds, not types (so that we can implicitly support definitions like `type MyInt64 int64`).

Reply:
Good point!